library/common/network/configurator.cc

```diff
@@ -229,9 +236,15 @@ void Configurator::refreshDns(envoy_netconf_t configuration_key, bool drain_conn
 
   if (auto dns_cache = dnsCache()) {
     ENVOY_LOG_EVENT(debug, "netconf_refresh_dns", std::to_string(configuration_key));
-    pending_drain_ = drain_connections && enable_drain_post_dns_refresh_;
-    // TODO(goaway): capture the list of currently cached hosts here in a set, to be checked
-    // for draining when DNS resolutions occur.
+
+    if (drain_connections && enable_drain_post_dns_refresh_) {
+      dns_cache->iterateHostMap(
+          [&](absl::string_view host,
+              const Extensions::Common::DynamicForwardProxy::DnsHostInfoSharedPtr&) {
+            hosts_to_drain_.emplace(host);
```

Comment:
should we explicitly empty `hosts_to_drain_` before we start adding hosts here? I think that we cannot end up in a state where we have dups in here but what if we had a host in the past that does not exist anymore (not sure whether this can happen)? 


Imagine we had DNS refresh 1) that included host x and y and DNS refresh 2) that included y host only. We want `hosts_to_drain_` to contain only `y` after the second DNS refresh is initialized

Reply:
`hosts_to_drain_` is a set, and should not contain dups. I don't think emptying the set in here or above actually closes the window on the small window for a race where a host ceases to exist between us enumerating hosts and then triggering the refresh.

I think putting a time bound on outstanding resolutions would, but that adds additional complexity, and given the narrow window, I'm not sure if it's really needed.